src/libmeasurement_kit/ffi.cpp

```diff
@@ -0,0 +1,99 @@
+// Part of Measurement Kit <https://measurement-kit.github.io/>.
+// Measurement Kit is free software under the BSD license. See AUTHORS
+// and LICENSE for more information on the copying conditions.
+
+#define MK_ENGINE_INTERNALS // we need to access engine internals
+
+#include <measurement_kit/ffi.h>
+
+#include <limits.h>
+#include <string.h>
+
+#include <exception>
+#include <new>
+#include <string>
+
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/engine.h>
+
+struct mk_event_ : public std::string {
+    using std::string::string;
+};
+
+static mk_event_t *mk_event_create(const nlohmann::json &json) noexcept {
+    return new (std::nothrow) mk_event_t{json.dump().data()};
+}
+
+const char *mk_event_serialize(mk_event_t *event) noexcept {
+    return (event) ? event->data() : nullptr;
+}
+
+void mk_event_destroy(mk_event_t *event) noexcept {
+    delete event; // handles nullptr
+}
+
+struct mk_task_ : mk::engine::Task {
+    using mk::engine::Task::Task;
+};
+
+mk_task_t *mk_task_start(const char *settings) noexcept {
+    return mk_task_start_ex(settings, nullptr, -1);
+}
+
+static void safe_copy(const char *reason, char *errbuf, size_t errbuf_siz) {
+    if (errbuf != nullptr && errbuf_siz > 0) {
+        size_t size = strlen(reason);
+        if (size > SIZE_MAX - 1) {
+            return;
+        }
+        size += 1; // account for the final '\0'
+        if (size < errbuf_siz) {
+            memcpy(errbuf, reason, size);
+            return;
+        }
+        // deal with truncation
+        memcpy(errbuf, reason, errbuf_siz);
+        errbuf[errbuf_siz - 1] = '\0'; // zero terminate
```

Comment:
Does this mean that if you have a target buffer that is smaller than what you copy into it, you truncate it to the maximum allowed length?

I think it's probably better to handle this as an unexpected behaviour and do some sort of error handling.

Reply:
Yes, I am truncating it. What sort of error handling do you suggest?

Reply:
I am not sure, but I think it's better to make it clear that we failed to write to the error buffer (maybe by writing in it "FAIL" or "ERR"), than truncating the error message (which is more likely to go undetected).

Reply:
@hellais I understand. So, for now I am going to document as a know limitation of this API that the error will be silently truncated in case of too-short buffer. I am however going to either open a small PR just to address that or a ticket. I do not think this issue is actually high priority, because in the end our primary use case is that of a JSON generated by a real JSON serializer and not manually, so I do not expect that there will be many cases where parsing fails.